pkgs/test_core/lib/src/direct_run.dart

```diff
@@ -0,0 +1,130 @@
+import 'package:test_api/src/backend/test.dart'; //ignore: implementation_imports
+import 'package:test_api/src/utils.dart'; // ignore: implementation_imports
+
+import 'runner/configuration.dart';
+import 'runner/engine.dart';
+import 'runner/plugin/environment.dart';
+import 'runner/reporter.dart';
+import 'runner/reporter/expanded.dart';
+import 'runner/runner_suite.dart';
+import 'runner/suite.dart';
+import 'util/print_sink.dart';
+
+/// Run all test cases declared in [testMain].
+///
+/// Test suite level metadata defined in annotations is not read. No filtering
+/// is applied except for the filtering defined by `solo` or `skip` arguments to
+/// `group` and `test`.
+Future<bool> directRunTests(FutureOr<void> Function() testMain,
+        {Reporter Function(Engine)? reporterFactory}) =>
+    _directRunTests(testMain, reporterFactory: reporterFactory);
+
+/// Run a single test declared in [testMain].
+///
+/// There must be exactly one test defined with the name [testName]. Note that
+/// not all tests and groups are checked, so a test case that may not be
+/// intended to be run (due to a `solo` on a different test) may still be run
+/// with this API. Only the test names returned by [enumerateTestCases] should
+/// be used.
+Future<bool> directRunSingleTest(
+        FutureOr<void> Function() testMain, String testName,
+        {Reporter Function(Engine)? reporterFactory}) =>
+    _directRunTests(testMain,
+        reporterFactory: reporterFactory, runSingleTestByName: testName);
+
+Future<bool> _directRunTests(FutureOr<void> Function() testMain,
+    {Reporter Function(Engine)? reporterFactory,
+    String? runSingleTestByName}) async {
+  reporterFactory ??= (engine) => ExpandedReporter.watch(engine, PrintSink(),
+      color: Configuration.empty.color, printPath: false, printPlatform: false);
+  final declarer = Declarer(filterToSingleTestName: runSingleTestByName);
+  await declarer.declare(testMain);
+
+  final suite = RunnerSuite(const PluginEnvironment(), SuiteConfiguration.empty,
+      declarer.build(), SuitePlatform(Runtime.vm, os: currentOSGuess),
+      path: p.prettyUri(Uri.base));
+
+  final engine = Engine()
+    ..suiteSink.add(suite)
+    ..suiteSink.close();
+
+  reporterFactory(engine);
+
+  final success = await runZoned(() => Invoker.guard(engine.run),
+      zoneValues: {#test.declarer: declarer});
+
+  if (runSingleTestByName != null) {
+    final testCount = engine.liveTests.length;
+    if (testCount > 1) {
+      throw DuplicateTestNameException(runSingleTestByName);
+    }
+    if (testCount == 0) {
+      throw MissingTestException(runSingleTestByName);
+    }
+  }
+  return success!;
+}
+
+/// Return the names of all tests declared by [testMain].
+///
+/// Test names declared must be unique. If any test repeats the name of a prior
+/// test a [DuplicateTestNameException] will be thrown.
+///
+/// Skipped tests are ignored.
+Future<Iterable<String>> enumerateTestCases(
```

Comment:
Should we type this as `Set` instead of `Iterable` - the contract is guaranteeing uniqueness so I think it makes sense and would avoid some potential copies for users who want a set.

Reply:
I'm uncertain when I want to used an ordered collection. As implemented the test names are returned in the same order that they'd be run (without randomization turned on). I don't know if that matters or not, or if it's better to return it as a `Set`. I'll try `Set` for now.